Various fixes, updates and tidy-up

tbooth opened the pull request:
I started off looking to fix bug 275, but this has turned into quite a big set of changes. The main points are:

- Splitting the tests out into categories and trying to make them a little neater
- Removing `setup.py` in favour of more modern `setup.cfg` and `pyproject.toml`
- Adding a test case for issue 275 and a fix (may not be the very best fix, but it passes the tests)
- Update of some docs
- Code cleanup. There is a large diff, but most changes are reformatting. I removed trailing whitespace from lines. I also changes out a bunch of old `%` string formatting for equivalent f-strings.

The tests are working nicely via GitHub actions but currently fail on Python 3.12 and 3.13. I think that there is a PR to fix this, so I'll look at this next.

I hope this PR will serve as a basis for ongoing maintenance of the codebase, but as you'll see I'm quite gung-ho with editing code that I think is "untidy". Please review my changes and *be direct* with any criticism!

tbooth commented:
The main functional change this PR makes to the code can be seen in this specific commit which comes in the middle of the other changes.

Fix for 275: https://github.com/bw2/ConfigArgParse/pull/302/commits/91120c53ef7b9ef0e37025044452ed7118555e8e

But while re-basing the changes I accidentally added the test for this to an earlier commit. In any case, the tests that cover this change are all in one new file: https://github.com/bw2/ConfigArgParse/blob/18dcb0c24f2836d7a053506c2935003d219d3d88/tests/test_issues.py

tbooth commented:
OK, @tristanlatr, so you're not even going to look at the code. @agyoungs, do you have any comments?

tbooth commented:
I've just updated my fork so that all the CI tests are passing. To do this I've had to mark a bunch of the tests that are copied from `test.test_argparse` as either `skip` or `expectedFailure`. The issues are:

1) `exit_on_error` is not implemented
2) `IntermixedArgs` is not implemented

I think these should be fairly easy, so I'm going to give that a go next. I'll do that in a separate branch so as to not add more commits to this PR.

agyoungs commented:
@tbooth Thanks for putting in some serious leg-work here! I was able to read through a fair amount of the code, but haven't done a deep dive yet. I don't think we should merge this all as-is, regardless. Some of my thoughts will echo @tristanlatr 's concerns, but that's not to say we shouldn't merge all these changes in eventually.

I'm going to make a big assumption in this paragraph, but I do think it's very important to get it right. My assumption is that @bw2 doesn't have the time to be the sole maintainer of this repo, but does want to be able to follow along with minimal effort with what's going on. That is going to be tricky to do with a PR titled "Various fixes, updates and tidy-up." I don't think this is just for @bw2's sake but the community's as well. I have a much easier time looking at open source projects where the title is a very clear description of the minimal required changes and tests to achieve the goal of the PR or issue. I'm assuming that @bw2 would like to see this approach based on his comment about establishing trust.

Okay, assumptions aside, here's my feedback as well. @tristanlatr suggested breaking this into multiple PR's and outlined at a high level a way to do that. I think that's a pretty reasonable starting point. On another issue, you mentioned code-formatting changes could be done with black. I definitely think any code formatting changes should be done by an automated tool (such as black) in a completely separate PR. This can be done first, last, or never, but I do think if we have multiple maintainers and really want community input it's going to be easier to solve all debates and just use a standard tool as you said. My only annoyance with black is it's a bit heavy handed, but it's trivial to integrate into a project and widely adopted so I'm all for it. Anyway, we definitely want that to be done in another PR entirely on its own.

I think the rest of your PR can still be further broken into other PRs (each line should probably be its own PR):
- Adding a test case for issue 275 and a fix (may not be the very best fix, but it passes the tests)
- Update of some docs; Removing setup.py in favour of more modern setup.cfg and pyproject.toml; Splitting the tests out into categories and trying to make them a little neater (could be its own separate PR as well honestly)
- Use black or other code formatting tool

And at the end of the day, none of us have any buy-in from @bw2 at this point. So I think whatever we can do to make his ownership of this repo easier to manage is going to benefit ConfigArgParse's community the most. The immediate next step is to get @bw2's buy-in, and I think we'll need to do that on the [maintenance issue](https://github.com/bw2/ConfigArgParse/issues/300). Obviously, any PRs that are ready to go for this process to be trialed will help that buy-in process

tbooth commented:
Thanks for the feedback. Hopefully @bw2 has time to comment soon.

In general I'd agree with you that lumping a bunch of unrelated commits into a big PR like this is not acceptable, but what you have here is all the updates from my own fork of the module representing a substantial amount of work. I could spend what is likely to be several hours of my time going back to square one and applying the changes in the order suggested by @tristanlatr, losing all my own commit history in the process, but at the end of the day you will still have the same amount of stuff to review.

Rather then fretting about the size of the PR as a whole, could you please comment specifically on two of the commits:

https://github.com/bw2/ConfigArgParse/pull/302/commits/9184df508253f3c77e5d8bf73855667e1a5ec07d which splits out the tests into separate files and introduces new tests in `tests/test_issues.py`.

https://github.com/bw2/ConfigArgParse/pull/302/commits/91120c53ef7b9ef0e37025044452ed7118555e8e which is my fix for the newly added tests.

Then we can worry about the other stuff later.

On the comment that changes to formatting are a low priority, I fundamentally disagree with this. I think that many codebases suffer from the mentality of leaving warts in the code for the sake of a minimal diff, and I'll generally prioritise making the existing code more legible before I try and fix a bug. This will lead to a PR that may have one or two reformatting commits prior to the actual fix. I think so long as the commits themselves are well structured this should not be a problem.

tbooth commented:
I've finished all that I plan to do on this code for now.

Since my last comment I've made the subclassing of config file parser types more simple and demonstrated the utility of that change in `tests/test_config_parser_subclasses.py`.

I've then pinned down and fixed some longstanding bugs. Test cases for all these are in `tests/test_issues.py`.

bw2 commented on setup.cfg:
```diff
@@ -0,0 +1,32 @@
+name = ConfigArgParse
+version = attr: setuptools_git_versioning.version_from_git
+author = Ben Weisburd
+author_email = weisburd@broadinistitute.org
```
would prefer not to list my email here

tbooth replied:
Sure, done. It can still be dug out of the Git history but was already there in the Git logs.

bw2 commented on configargparse.py:
```diff
@@ -119,6 +121,44 @@ def serialize(self, items):
+            else:
+                yield newkey, str(newvalue)
+
+    def tweak_value(self, key, value, filename):
```
function name: To me "tweak" implies small modification, but the example subclasses use this function in various other ways. I'd vote for changing this to something like `process_value` or `handle_config_entry` (since it takes both the key and the value). 
Also, the docs nicely explain what the return value can contain, but should also clarify what the filename arg is, and that extra key/values in the dict will be treated as if they were found in that config file (?) while keys with a None value will be treated as if they were absent from the config file(?)

tbooth replied:
Agreed. Maybe `amend_config_entry` is a fitting name?

bw2 replied:
how about `process_config_entry` since this method may be useful for scenarios that don't modify the entry?

bw2 commented on configargparse.py:
```diff
@@ -119,6 +121,44 @@ def serialize(self, items):
         """
         raise NotImplementedError("serialize(..) not implemented")
 
+    def _tweak_value(self, key, value, filename):
```
Having two two methods with the same name except for the _ prefix is not something I've seen before. Unless this is some naming convention I'm not aware of, I'd vote for more a more explicit name difference (eg. these are not the best suggestions, but perhaps _tweak_value_base or alternatively .tweak_value_impl)

tbooth replied:
This is used for `parse_known_args`, `_parse_known_args2` and `_parse_known_args` in the [underlying class](https://github.com/python/cpython/blob/main/Lib/argparse.py) but I'd be happy to rename this pseudo-private function to something more explicit.

bw2 commented on configargparse.py:
```diff
@@ -813,19 +927,68 @@ def parse_known_args(
+                                          env_vars=env_vars,
+                                          ignore_help_args=ignore_help_args)
+
+    def _parse_known_args_cap(
```
Not sure what '_cap' suffix stands for - maybe _parse_known_args_impl is better name?

tbooth replied:
cap = ConfigArgParse since this function wraps the corresponding function from the base class, rather than overriding it. I agree that if not renamed this should at least be explicit in the docstring.

bw2 replied:
would prefer replacing `_cap` with `_impl` or `_base` or something

bw2 commented on configargparse.py:
```diff
@@ -813,19 +927,68 @@ def parse_known_args(
+        Arguments:
+            args: a list of args as in argparse, or a string (eg. "-x blah -y blaz")
+            config_file_contents (str). Used for testing.
+            env_vars (dict). Used for testing.
```
namespace arg not documented

tbooth replied:
True, but not my bug ;-)

bw2 commented on configargparse.py:
```diff
@@ -813,19 +927,68 @@ def parse_known_args(
+                                          env_vars=env_vars,
+                                          ignore_help_args=ignore_help_args)
+
+    def parse_known_intermixed_args(
```
sorry, naive question until I have more time to dig into the differences - why are two separate sets of methods needed for intermixed and non-intermixed args? Why can't all cases be handled as intermixed?

tbooth replied:
A very good question, and TBH I'm not entirely sure of the full details. I followed the underlying `argparse.py` code and just made sure the upstream tests were working if I call the corresponding function in the superclass. If I force everything to be intermixed then I get a *lot* of test failures (from `python3 -munittest tests.test_argparse_tests`). The [upstream docs](https://docs.python.org/3/library/argparse.html#intermixed-parsing) say, amongst other things:

> These \[intermixed\] parsers do not support all the argparse features, and will raise exceptions if unsupported 
> features are used. In particular, subparsers, and mutually exclusive groups that include both optionals and 
> positionals are not supported.

bw2 commented:
I left some initial comments. I agree that, going forward, smaller / more focused PRs are better. For now, let's see how far we can get with reviewing the current version and where the bulk of the comments will be.. perhaps then it'll be more obvious if/how splitting is needed.

tbooth commented:
> I left some initial comments. I agree that, going forward, smaller / more focused PRs are better. For now, let's see how far we can get with reviewing the current version and where the bulk of the comments will be.. perhaps then it'll be more obvious if/how splitting is needed.

Thanks for making the time to look through this. Obviously it would be very easy for me to make a smaller PR by taking just the first few commits into a new branch on my fork, but re-basing my later commits directly onto the upstream master branch is more of a PITA so I'll only do that if I really need to. The more systematic changes that I made are:

- [ ] Code reformatting and tidy. With hindsight I should not have done this manually as it bloats the diff, but I went back later and re-based all my other changes on **black**-ened code, so if you are happy to have the code formatted that way then applying **black** should be the first order of business (I can make a PR, but you can also very easily do it for yourself).
- [ ] I split out the tests. This just seemed a no-brainer to me as the test file was getting very large. Yes, the underlying module has a single big test file, but that lives in the CPython code and I guess they want to keep a lid on the number of source files.
- [ ] I removed use of `%` string formatting as I find it hard to read, and while it's not quite deprecated it's basically a Python 2 relic. Having a mixture of `%`-formats along with `.format()` and *f-strings* definitely seems wrong. On the other hand, the underlying `argparse.py` module is still chock full of this old syntax.

If these changes are all acceptable, then being a bit more selective about the other stuff I changed should be fine, without making a huge amount of work to cherry pick the commits. Of course, I hope all my changes are good, and many are simply necessary to maintain test-suite compatibility with the newer `argparse`.

tbooth commented:
For review purposes, here are other significant things I changed:

- [ ] All tests pass on all Python versions suppoted by GitHub actions
- [ ] Including support for `parse_intermixed_args` (tests were previously bypassed)
- [ ] Move from `setup.py` to `setup.cfg`
- [ ] Use of `setuptools-git-versioning`
- [ ] New hook for subclassing, and sample code to demonstrate (name of hook method TBD)
- [ ] Fix for bug regarding '--'
- [ ] Fix for bug recognising grouped short args (eg. `-xyz`)
- [ ] Addressed some other bugs (see the issues test cases)

Other notes:

- Fully consistent behaviour with sub-parsers is not supported, and cannot be within the current design.

Issues we may want to report to upstream (for `argparse`)

- I believe the current `argparse` behaviour with `--` is broken when prefix char is not `-`
- I believe the current `BooleanOptionalAction` is broken in that it ignores custom prefix char setting

kingbuzzman commented:
There is a lot of great stuff in here. But this is a lot of absorb and review, please break this into smaller PRs that are focused on a single purpose.
